src/pybind/mgr/dashboard/services/oauth2.py

```diff
@@ -0,0 +1,132 @@
+import json
+import logging
+from typing import Dict, List
+from urllib.parse import quote
+
+import cherrypy
+import requests
+from cherrypy import _ThreadLocalProxy
+
+from dashboard.controllers.auth_classes import decode_jwt_segment
+
+from .. import mgr
+from ..services.access_control import Role, User, UserAlreadyExists
+from ..tools import prepare_url_prefix
+
+logger = logging.getLogger('OAuth2')
+
+
+class OAuth2:
+    @classmethod
+    def get_token(cls, request: _ThreadLocalProxy) -> str:
+        try:
+            return request.cookie['token'].value
+        except KeyError:
+            return request.headers.get('X-Access-Token')
+
+    @classmethod
+    def set_token(cls, token: str):
+        cherrypy.request.jwt = token
+        cherrypy.request.jwt_payload = cls.get_token_payload()
+        cherrypy.request.user = cls.get_user(token)
+
+    @classmethod
+    def get_token_payload(cls) -> Dict[str, str]:
+        try:
+            return cherrypy.request.jwt_payload
+        except AttributeError:
+            pass
+        try:
+            return decode_jwt_segment(cherrypy.request.jwt.split(".")[1])
+        except AttributeError:
+            return {}
+
+    @classmethod
+    def set_token_payload(cls, token):
+        cherrypy.request.jwt_payload = decode_jwt_segment(token.split(".")[1])
+
+    @classmethod
+    def get_user_roles(cls):
+        roles: List[Role] = []
+        user_roles: List[Role] = []
+        try:
+            jwt_payload = cherrypy.request.jwt_payload
+        except AttributeError:
+            raise cherrypy.HTTPError()
+
+        # check for client roes
+        if 'resource_access' in jwt_payload:
+            # Find the first value where the key is not 'account'
+            roles = next((value['roles'] for key, value in jwt_payload['resource_access'].items()
+                          if key != "account"), user_roles)
+        # check for global roles
+        elif 'realm_access' in jwt_payload:
+            roles = next((value['roles'] for _, value in jwt_payload['realm_access'].items()),
+                         user_roles)
+        else:
+            logger.info('No user roles found, continuing as read-only user')
```

Comment:
If no role is found, it might be the case that this user shouldn't be allowed any access to Ceph Dashboard. Keep in mind that SSOs in larger companies/users, include all employee accounts, and in many cases only a small subset of them should have access to IT systems (including Ceph).

We can make this more flexible by adding some module option to tune this behaviour (e.g.: `sso_role_fallback` = "deny" or "read-only").

Reply:
> We can make this more flexible by adding some module option to tune this behaviour (e.g.: sso_role_fallback = "deny" or "read-only").

That sounds cool, I'll think about that.

> If no role is found, it might be the case that this user shouldn't be allowed any access to Ceph Dashboard.
 
Yeah, I get that. 
My thought process here was that if you've made it this far you have properly configured the `oauth2-proxy` service with your IdP's `client_ID `and `secret`, so only users of that spcified client should be able to get to this part of the code.
Therefore I thought it would be easier for a user that expected to have X permissions to realize that those were not set. Rather than simply not logging in, since well... everything was in place to be logged in!
Also I thought giving read-only access was harmless and therefore I went with it.

I'm not really sure if I should just remove this, as it is true that if you have not properly configured your user, maybe it just shouldn't be able to login.

So your call @epuertat , or anyone who might read all this... Should I remove the default read-only user permissions if no permissions are found in place for the user?

Reply:
@nizamial09 @rkachach @afreen23 @rkachach we summon you!

IMHO this is the issue with authentication vs authorization. All SSO organization users should be valid users to the Ceph Dashboard, but only a few should be authorized. In any case, my proposal would allow for both scenarios, being the "reject" fallback the default based on the [Principle of Least Authority](https://en.wikipedia.org/wiki/Principle_of_least_privilege).

Reply:
In the end I went with the `Principle of Least Authority` as you mentioned, and got rid of the default `read-only` role.
Now if no roles are provided, user won't be able to access

Reply:
@Pegonzal @epuertat my vote would be for not letting the user in. "readonly" sounds harmless but it is already giving access to the system and a malicious will have access to see the system configuration and what's going on (maybe preparing for his next attack). IMHO that would be a serious security breach and I'd recommend just refusing the access. If this user really needs to login he will reach out his IT and they will help him to investigate what's going on. My 2 cents.